devices/webcam/hand_tracking_utility.py

```diff
@@ -0,0 +1,600 @@
+    
+    NormalizedLandmarkList:
+        @attributes:
+            landmark: RepeatedCompositeContainer, list-like container with up to 21 hand pose landmarks with their normalized (X, Y, Z) coordinates as attributes
+    """
+
+    frame: np.ndarray
+    landmarks: List[NormalizedLandmarkList]
+
+class PoseEstimationConfig(lg.Config):
+    """
+    Configuration for MediaPipe's hand pose esitmation graph
+    See https://google.github.io/mediapipe/solutions/hands.html
+
+    @attributes:
+        model_complexity: int
+        min_detection_confidence: float
+        min_tracking_confidence: float
+    """
+
+    model_complexity: int
+    min_detection_confidence: float
+    min_tracking_confidence: float
+
+class PoseEstimationState(lg.State):
+    """
+    State object for hand pose estimation node
+    See https://google.github.io/mediapipe/solutions/hands.html
+
+    @attributes:
+        hands: mediapipe.python.solutions.hands.Hands
+    """
+
+    hands: Optional[HandsType.Hands] = None
+
+class PoseEstimation(lg.Node):
+    """
+    PoseEstimation node, runs VideoFrame streams through MediaPipe's Hands graph to generate pose data
+
+    @topics:
+        INPUT: VideoFrame
+        OUTPUT: PoseFrame
+    
+    @attributes:
+        config: PoseEstimationConfig
+        state: PoseEstimationState
+    """
+
+    INPUT = lg.Topic(VideoFrame)
+    OUTPUT = lg.Topic(PoseFrame)
+    config: PoseEstimationConfig
+    state: PoseEstimationState
+
+    @lg.subscriber(INPUT)
+    @lg.publisher(OUTPUT)
+    async def processFrame(self, message: VideoFrame) -> lg.AsyncPublisher:
+        results: NamedTuple = self.state.hands.process(cv2.cvtColor(message.frame, cv2.COLOR_BGR2RGB))
+        landmarks = results.multi_hand_landmarks
+        if landmarks is None:
+            landmarks = []
+        yield self.OUTPUT, PoseFrame(landmarks = landmarks, frame = message.frame)
+
+    def setup(self):
+        self.state.hands = mp_hands.Hands(
+            model_complexity = self.config.model_complexity,
+            min_detection_confidence = self.config.min_detection_confidence,
+            min_tracking_confidence = self.config.min_tracking_confidence)
+
+class LoggingConfig(lg.Config):
+    """
+    Configuration for the data logging node
+
+    @attributes:
+        log_images: bool, whether or not to log images to file
+        log_poses: bool, whether or not to log pose data to file
+        file_path: str, the path of the log file
+    """
+
+    log_images: bool
+    log_poses: bool
+    file_path: str
+
+class LoggingState(lg.State):
+    """
+    State object for data logging
+
+    @attributes:
+        file: h5py.File
+        closed: bool, true if the file is closed
+        frame_index: the current index of the frame received from the VideoFrame topic
+    """
+
+    file: Optional[h5py.File] = None
+    closed: bool = True
+    frame_index: int = 0
+
+class Logging(lg.Node):
```

Comment:
LabGraph HDF5 logger shall be used instead of a standalone logging node so that the logging behavior could be optimized at a central location.

Reply:
I have LabGraph's HDF5 logger working and producing files, but I'm having trouble understanding how the data is converted to numpy arrays and how one would go about reconstructing objects from them. Any pointers?

Reply:
An example can be found [here](https://github.com/facebookresearch/labgraph/issues/20) (step 6).

Reply:
Excellent, thanks!

Reply:
This solution now uses LabGraph's HDF5 logger and replayer.